set.contains (with lowering) + piecewise Presburger expression

Dinistro opened the pull request:
This PR contains set.contains and its lowering pass. Furthermore it introduces Piecewise Presburger expressions into the dialect. This includes a data structure, an operation (with attribute and type), parsing, printing and some minor refactoring.

PS: Sorry for mixing different feature in one PR...

Dinistro committed:
feature: presburger.contains operation

Dinistro committed:
feature: presburger.contains lowering to std

This commit introduces a presburger to standrad lowering pass that can
be enabled in mlir-opt with --convert-presburger-to-std.

At the moment it only supports lowering of presburger.contains.

Dinistro committed:
feature: add verification to contains

Dinistro committed:
refactoring: add set prefix to set attribute

Dinistro committed:
feature: piecewise Presburger expression

Dinistro commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -260,12 +267,51 @@ class SetExpr : public Expr {
+  std::unique_ptr<Expr> constraints;
+};
+
+class PwExprExpr : public Expr {
```
I'm not really happy with the structure of this class, but I didn't find a nicer way yet.

Dinistro commented:
I will work on some comments for the parsing infrastructure. Some parts of if include hacks to enable custom parsing and the error handling seems to be a bit arbitrary (especially locations of messages).

Dinistro commented on mlir/lib/Dialect/Presburger/PresburgerDialect.cpp:
```diff
@@ -28,29 +28,58 @@ PresburgerDialect::PresburgerDialect(mlir::MLIRContext *context)
-  }
+  // Parse the kind keyword first.
+  std::unique_ptr<VariableExpr> attrKind;
+  if (failed(p.parseVariable(attrKind)))
```
I do not like this hack, especially how it uses parseVariable, even though it parsers a special kind of keyword: `set` or `pwExpr`. I think this could be integrated deeper into the parser.

Dinistro commented on mlir/lib/Dialect/Presburger/PresburgerDialect.cpp:
```diff
@@ -66,14 +95,28 @@ Type parsePresburgerSetType(DialectAsmParser &parser, MLIRContext *context) {
 
+Type parsePresburgerPwExprType(DialectAsmParser &parser, MLIRContext *context) {
+  unsigned dimCount, symbolCount;
+  if (parser.parseLess() || parser.parseInteger<unsigned>(dimCount) ||
```
This could be a helper function, as it is used for both types

superty replied:
Agreed.

Dinistro commented on mlir/lib/Dialect/Presburger/PresburgerOps.cpp:
```diff
@@ -49,6 +49,31 @@ static LogicalResult verify(SetOp op) {
+}
+
+static LogicalResult verify(PwExprOp op) {
+  // TODO do we need something here?
```
We should theoretically check if the domains are disjunct somewhere. I'm just not sure if this is too expensive

superty replied:
Need to decide whether domains can be non-disjoint if the value of the expression is equal in the intersection of the two domain pieces.

Dinistro replied:
We could keep it for now without verification and discuss this in the next meeting. Or do you, @tobiasgrosser , have a strong opinion on this one. 
Note that we also need to specify what happens if the domains do not cover the complete space, i.e. what the expression should evaluate to in such cases.

tobiasgrosser replied:
I don't have a strong opinion. ;-)

Dinistro commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -63,6 +63,10 @@ std::string Token::name(Token::Kind kind) {
     return "\"and\"";
   case Token::Kind::Or:
     return "\"or\"";
+  case Token::Kind::Arrow:
```
I think that this method can currently read too far, i.e. out of the space of the provided `buffer`. Should this be forbidden or is this a problem for the owner of the `Lexer`?

superty replied:
I don't see what you mean. (Has this issue been resolved already?)

Dinistro replied:
There is no check if `curPtr` points further than the `StringRef`. Now that I think of it, the custom parser will only be called on the string inside `<"` `">` and therefore will always throw an error if it reaches `"` as this is an unexpected character.

superty replied:
I don't understand where `curPtr` is accessed in `Token::name`?

Dinistro replied:
Oh, turns out that I commented on the wrong method, sorry. It should have been in `Lexer::nextToken`, sorry.

superty replied:
I feel it would be better to handle this case explicitly.

Dinistro replied:
I'm not sure if this can be done without rewriting bigger parts of the `Lexer`. `Lexer::nextToken` returns a `Token` and changing this to `LogicalResult` might be cumbersome. Everything would need to be changed for a case that cannot happen when the `PresburgerParser` is correctly used (it is only supposed to be used as a custom Attribute parser). The "correct usage" should, of course, be documented somewhere.  

Note:
Explicitly enforcing that `curPtr` is only dereferenced within the bounds of `buffer` would require checks all over the place, i.e. either on every increment or on every dereferencing.

superty replied:
Sorry, I should've written more clearly. IMO, an assert failure here in case of out of bounds is sufficient, no need to return a `LogicalResult`.

I see your point about checks all over the place. However, I just noticed that `ArrayRef`, and hence `StringRef`, assert the index to be within bounds in the `operator[]`. Therefore, switching to indices instead of pointers should resolve this transparently.

Dinistro replied:
No worries. 
Nice to know that they use assertions. I will change it then.

Dinistro replied:
`Lexer::next` always calls `Lexer::nextToken`. To make sure we do not run into an out of bounds assertion, I'll add a check and return an EOF token.

Dinistro committed:
refactoring: comments for parser and small changes

superty commented on mlir/include/mlir/Analysis/Presburger/PwExpr.h:
```diff
@@ -0,0 +1,33 @@
+#include "mlir/Analysis/Presburger/Set.h"
+
+namespace mlir {
+class PresburgerPwExpr {
```
Hm, the salient part is that the expression is piecewise _affine_. What about a name like PiecewiseAffine or PWAffine?

superty replied:
For acronyms like pw, I'm not sure if it should be Pw or PW. "Pw" isn't really one word. Is there precedent in the MLIR codebase?

Dinistro replied:
In the end this should not only support affine expressions, but all kinds of Presburger expressions. So they should be called piecewise Presburger expressions, but I was not sure if this name would be long.

Dinistro replied:
> Is there precedent in the MLIR codebase? 

Not that I know of.

superty replied:
What kinds of other presburger expressions? Would that functionality be incorporated into the same class?

Dinistro replied:
Divisions for example. But I'm not sure if this will be possible in my remaining time.

tobiasgrosser replied:
Even divisions are still Quasi-Affine.

The reason why isl calls these expressions affine is that these expressions are modelled as explicit functions rather than relations encoded through a list of constraints.

Dinistro replied:
Would it be better to call them piecewise affine expressions then? I assume that piecewise quasi-affine expression might be too long.

Any suggestions on how to name them in the dialect? `PwAffineExpression`, `PwAffineExpr`, `PwQuasiAffineExpr`, ... ?

tobiasgrosser replied:
Think about this, I think `PresburgerExpr` could make a lot of sense, as the term 'Expr' stands for the fact that this is an explicit expression (rather than a collection of constraints), and Presburger could be a shorthand for 'PiecewiseQuasiAffine'.

tobiasgrosser replied:
Alternatively, following isl more closely, we could use `PiecewiseAffineExpr`. This one might be more helpful if we want to distinguish them from AffineExpr. I personally would not try to make these terms overly short.

Dinistro replied:
I like the `PiecewiseAffineExpr` the most, as it highlights the "piecewise" property of the construct. 

We could also go for `PiecewisePresburgerExpr` as this would be the exactest naming. As the domains are specified by Presburger sets the other option could be a little misleading.

Dinistro replied:
This would also blend in "nicely" with the corresponding operation: `presburger.piecewiseExpr ...`

tobiasgrosser replied:
I think `PiecewisePresburgerExpr` does not really make a lot of sense. Presburger expressions can have many pieces, so adding pieces on top does not add anything.

Dinistro replied:
Good point, taken this into consideration `PresburgerExpr` would be good. Then we can use `presburger.expr` as the defining operation.

superty commented on mlir/include/mlir/Analysis/Presburger/PwExpr.h:
```diff
@@ -0,0 +1,33 @@
+#include "mlir/Analysis/Presburger/Set.h"
+
+namespace mlir {
+class PresburgerPwExpr {
```
I think it makes sense at this point to put this in the `presburger` namespace? Not sure if it makes sense to have the dialect and the "math" in the same namespace though. One of them has to be given a different name. (I would also change my `PresburgerBasicSet`to strip the "Presburger" prefix and include it in the "math" namespace).

Dinistro replied:
That would make sense. There is no other `Analysis` subfolder, therefore we need to come up with a naming scheme ourselves.

Dinistro replied:
Do you have any suggestions?

superty replied:
I'm not sure. Files in Analysis don't seem to have a namespace, but maybe we could make an `analysis::presburger` anyway, even if there is nothing else in `analysis` but `presburger`. This seems cleaner than coming up with a strange name for one of them, unless we append `dialect` to the namespace for dialects, but this would be a bit strange when no other dialect does this.

Dinistro replied:
Agreed, I will add the `analysis::presburger` namespace

superty commented on mlir/test/Conversion/PresburgerToStandard/convert-presburger-to-standard.mlir:
```diff
@@ -0,0 +1,56 @@
+// CHECK-NEXT:   %25 = or %false, %24 : i1
+// CHECK-NEXT:   return %25 : i1
+// CHECK-NEXT: }
+func @contains_runtime(%x0 : index, %x1 : index, %s0 : index) -> i1 {
```
Not very familiar with MLIR itself, but what does this test do? Are you sure it is not too brittle?

Dinistro replied:
To be honest, I did not know how else I could test this. Perhaps @tobiasgrosser has a better idea

tobiasgrosser replied:
If we control the lowering, then this seems fine to me.

Dinistro replied:
Then we keep it like this for now. I do not know of a better way yet.

superty commented on mlir/lib/Conversion/PresburgerToStandard/PresburgerToStandard.cpp:
```diff
@@ -0,0 +1,130 @@
+
+void mlir::populatePresburgerToStdConversionPatterns(
+    OwningRewritePatternList &patterns, MLIRContext *ctx) {
+  // clang-format off
```
I don't understand, why do we need to turn formatting off here?

Dinistro replied:
This was done in at most other points where patterns get inserted. If there are multiple patterns, they can be written under each other, but clang-format squashes them in separate lines. 

This will be relevant as soon ad the pwExpr lowering is in place.

superty commented on mlir/lib/Conversion/PresburgerToStandard/PresburgerToStandard.cpp:
```diff
@@ -0,0 +1,130 @@
+using namespace mlir::presburger;
+
+namespace {
+struct PresburgerSetTransformer {
```
Some basic documentation here would be nice.

superty commented on mlir/lib/Analysis/Presburger/PwExpr.cpp:
```diff
@@ -0,0 +1,96 @@
+  os << ")";
+}
+
+// TODO somehow merge this with PresburgerSet
```
Maybe a `PresburgerPrinter` class?

Dinistro replied:
Thought about that, but wasn't sure. I will try to implement this

superty commented on mlir/lib/Analysis/Presburger/PwExpr.cpp:
```diff
@@ -0,0 +1,96 @@
+
+void PresburgerPwExpr::printVar(raw_ostream &os, int64_t val, unsigned i,
+                                unsigned &countNonZero) const {
+  if (val == 0) {
```
I feel that instead of passing around countNonZero, it might make more sense to just have the first part of the function where we output the coefficient in printExpr and only keep the last part where we print the variable in this function.

Dinistro replied:
I will move all this into a PresburgerPrinter class and see what makes sense there

superty commented on mlir/lib/Analysis/Presburger/PwExpr.cpp:
```diff
@@ -0,0 +1,96 @@
+#include "mlir/Analysis/Presburger/PwExpr.h"
+
+using namespace mlir;
```
Some documentation would be nice. In particular, an example of how the printed output looks like for the print functions will help the reader understand what the code is doing.

Dinistro replied:
Good point, I will work on this

superty commented on mlir/lib/Analysis/Presburger/PwExpr.cpp:
```diff
@@ -0,0 +1,96 @@
+
+// TODO somehow merge this with PresburgerSet
+void PresburgerPwExpr::printVariableList(raw_ostream &os) const {
+  assert(domains.size() > 0 && "pwExpr needs atleast one piece");
```
`!domains.empty()` might be cleaner

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
+  /// Returns true if the token is of Kind kind.
   bool isa(Kind kind);
+
+  /// Retruns the StringRef corresponding to the content.
```
nit: "Returns"

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
     And,
     Or,
+    Arrow,
+    SemiColon,
```
nit: `Semicolon` is one word.

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -25,6 +26,11 @@ namespace mlir {
 namespace presburger {
```
The TODO and `#include <memory>` a few lines above can be removed now. (github doesn't let me comment on that line directly in the review, probably as it's not part of the diff)

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
 };
 
+/// Splits the buffer into Tokens and is able to report errors at certain
+/// location while using the provided ErrorCallback.
```
nit: "certain locations by using"?

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
+  /// Returns the next token and consumes it.
   Token next();
+
+  /// Consumes a token with specified kind. If non is present it is a no-op.
```
"If non is present"

Shouldn't it be "if the next token is not of the specified kind"?

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
+  /// Creates an integer token while consuming all digits
   Token integer(const char *start);
+
+  /// Create an identifier of keyword. An identifier has to start witha n
```
nits: identifier *or* keyword.

start *with an*

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
   bool isAlpha(char c);
+
+  /// Returns a Token from start to curPtr
   Token atom(Token::Kind kind, const char *start);
```
Something like `getAtom` might be a better name for a function.

superty replied:
Would passing around indices be better than passing raw pointers?

Dinistro replied:
> Something like getAtom might be a better name for a function.

Agreed, I'll change this.

> Would passing around indices be better than passing raw pointers?

If we have the pointer we do not need an additional lookup of a starting address. The `StringRef` of that the lexer is using directly points into the file buffer and should be valid until this file is fully compiled. Therefore the pointer will be valid. 
The implementation itself uses `curPtr` to determine the size of the produced `StringRef`.

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
+  /// Returns a Token from start to curPtr
   Token atom(Token::Kind kind, const char *start);
+
+  /// Creates an integer token while consuming all digits
```
Would `consumeInteger` be a better name?

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
+
+  /// Create an identifier of keyword. An identifier has to start witha n
+  /// alphabetic char and after that contains a sequence of alphanumeric chars.
   Token identifierOrKeyword(const char *start);
```
Would `consumeIdentifierOrKeyword` be more descriptive?

superty commented on mlir/include/mlir/Dialect/Presburger/Parser.h:
```diff
@@ -51,51 +57,94 @@ class Token {
+
+  /// Consumes the next token and emits an error if it doesn't match the
+  /// provided kind.
   LogicalResult nextAssertKind(Token::Kind kind);
```
Should the name be changed since we now don't assert per se but emit an error?

Dinistro replied:
Something like `consumeOrError` or `consumeOrEmitError`?

Dinistro replied:
I will go for `consumeKindOrError`

superty commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -210,26 +233,50 @@ LogicalResult Lexer::nextAssertKind(Token::Kind kind, Token &token) {
 
 //===----------------------------------------------------------------------===//
-// Parser (from libint)
+// Parser (partly from libint)
```
We can get rid of the attribution to libint, unless there's a particular reason you want to keep it around

Dinistro replied:
Now that I added some documentation, it can be dropped.

superty commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -635,9 +827,14 @@ PresburgerSetParser::parseSum(Expr *expr,
 }
 
+/// Takes a TermExpr and addapts the matchin coefficient or the constant to this
+/// terms value. To determine the coefficient id it luuks it up in the
```
nit: *looks* it up

superty commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -635,9 +827,14 @@ PresburgerSetParser::parseSum(Expr *expr,
+/// terms value. To determine the coefficient id it luuks it up in the
+/// nameToIndex mappings
+///
+/// Failes if the variable name is unknown
```
nit: *Fails* if the...

superty commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -460,6 +620,10 @@ LogicalResult Parser::parseVariable(std::unique_ptr<VariableExpr> &vExpr) {
   return success();
 }
 
+/// Parse an signless integer.
```
nit: *a* signless integer

superty commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -452,6 +608,10 @@ LogicalResult Parser::parseTerm(std::unique_ptr<TermExpr> &term,
   return success();
 }
 
+/// Parse an variable.
```
nit: *a* variable

superty commented on mlir/lib/Dialect/Presburger/Parser.cpp:
```diff
@@ -365,6 +471,47 @@ LogicalResult Parser::parseAnd(std::unique_ptr<Expr> &expr) {
+LogicalResult
+Parser::parsePieces(SmallVector<std::unique_ptr<PieceExpr>, 4> &pieces) {
+  do {
+    lexer.next();
```
What is this supposed to consume? The semicolon between pieces? But then why does the first run of the loop also do this? Please clarify with a comment, and mention the delimiter between pieces in the documentation at the top

Dinistro replied:
This is my bad. The calling method doesn't consume the previous token, therefore this has to be done here. I did not notice (or ignored it), as this made the implementation easier. 

Furthermore it seems that the comment is wrong. This consumes a list of pieces, not only one.

Dinistro replied:
Changed it to use similar concepts as `parseAnd` and `parseOr`

superty reviewed (commented):
I left mostly code style-related comments and suggestions.

Also, all the copyright notices should be changed to the MLIR-style ones.

Dinistro committed:
refactoring: applied arjun's suggestions

Dinistro committed:
feature: PresburgerPrinter class

Added a Printer class that prints Presburger sets and piecewise
Presburger expressions. This showed a small bug that required test case
changes.

Dinistro commented:
I added a PresburgerPrinter that handles printing of sets an piecewise expressions. Currently all methods are static, therefore a class is perhaps not the best idea. Should I just create functions for this? 

I had to move it into the `Analysis` directory as PresburgerSet::dump and PwExpr::dump depend on it.

Dinistro committed:
refactoring: implemented the remaining feedback

This includes parser function renaming, general commenting and a
small refactoring in the PresburgerDialect

Dinistro commented:
> Also, all the copyright notices should be changed to the MLIR-style ones.

This wasn't done in any of my created files so far. I suggest that I do this at the end of the thesis with just one commit for all the files.

Dinistro committed:
refactoring: renaming PresburgerPwExpr to PresburgerExpr

This includes pwExpr to expr in the dialect.

Superty commented:
> I added a PresburgerPrinter that handles printing of sets an piecewise expressions. Currently all methods are static, therefore a class is perhaps not the best idea. Should I just create functions for this?

I don't have a strong opinion. Just creating functions seems fine.

superty commented on mlir/lib/Conversion/PresburgerToStandard/PresburgerToStandard.cpp:
```diff
@@ -12,12 +12,23 @@ using namespace mlir;
                            Location loc)
       : builder(builder), dimAndSymbolValues(dimAndSymbolValues), loc(loc) {}
 
+  /// This create a sequence of std operations that check if a provided point
```
nit: create -> creates

superty commented on mlir/lib/Conversion/PresburgerToStandard/PresburgerToStandard.cpp:
```diff
@@ -80,6 +91,9 @@ struct PresburgerSetTransformer {
 } // namespace
 
 /// Presburger contains are replaced by runtime checks
+// TODO this should perhaps be matched on a complete function, as it otherwise
```
nit: Triple slash for top-level comments

superty commented on mlir/include/mlir/Analysis/Presburger/Expr.h:
```diff
@@ -0,0 +1,48 @@
+#include "mlir/Analysis/Presburger/Set.h"
+
+namespace mlir {
+/// This class implements a data structure that represents a piecewise
```
Presburger is already piecewise -- maybe something like "represents a Presburger expression, i.e., a piecewise affine expression"

superty commented on mlir/include/mlir/Analysis/Presburger/Expr.h:
```diff
@@ -0,0 +1,48 @@
+/// This class implements a data structure that represents a piecewise
+/// Presburger expression. Each piece is define by a domain and an expression
+/// that is applied to points lying in this domain. Domains are represented as
+/// parameteric Presburger sets, expressions by normal Presburger expressions.
```
Expressions are affine or quasi-affine, but cannot be piecewise as Presburger expressions can?

superty commented on mlir/include/mlir/Analysis/Presburger/Expr.h:
```diff
@@ -0,0 +1,48 @@
+
+private:
+  // print helpers
+  void printVariableList(raw_ostream &os) const;
```
These print helpers aren't defined anymore, right?

Dinistro replied:
Good catch, thanks.

superty commented on mlir/lib/Analysis/Presburger/Printer.cpp:
```diff
@@ -0,0 +1,169 @@
+  }
+}
+
+/// Prints a Presburger expression. `coeffs` contains all the coefficients:
```
I feel the documentation can be slightly changed to be more specific: this prints an affine expression (Presburger expressions could be piecewise)

superty commented on mlir/lib/Analysis/Presburger/Printer.cpp:
```diff
@@ -0,0 +1,169 @@
+///
+/// Returns failure if the coefficient value is 0 and therefore is not printed.
+///
+LogicalResult PresburgerPrinter::printCoef(raw_ostream &os, int64_t val,
```
nit: I feel `coeff` is a more standard short form for "coefficient"

superty replied:
I feel that not printing zero coefficients does not really indicate a _failure_ per se, hence using a bool may be more appropriate for the return type.

Dinistro replied:
I wasn't sure how they prefer it in LLVM/MLIR,i thought that it was somewhere stated that boolean returns can be confusing. But in this case I agree.

Dinistro commented:
> > I added a PresburgerPrinter that handles printing of sets an piecewise expressions. Currently all methods are static, therefore a class is perhaps not the best idea. Should I just create functions for this?
> 
> I don't have a strong opinion. Just creating functions seems fine.

Agreed. It was just the natural way for me as I have a Java background.

Dinistro committed:
refactoring: implementing arjuns suggestions

Dinistro committed:
refactoring: add analysis::presburger namespace

Dinistro committed:
fix: Lexer respecting buffer bounds

Superty commented:
LGTM

Dinistro commented:
@tobiasgrosser Do you have something that should be changed? If not we can merge this, I think.

Dinistro committed:
refactoring: removing last pwExpr references

Dinistro committed:
fix: Presburger expressions with no constraints

Dinistro commented:
Fixed the issue and removed the last references to "PwExpr" and "piecewise Presburger expressions". Now everything should be working as expected.
